### PR TITLE
[Improvement](sort) fallback to partial sort small block if topN is small

### PR DESCRIPTION
**Pull request by Gabriel39**
When `limit` is small in topN node and some string columns need to be materialized, `Block::merge` brings high overhead when accumulating big blocks. 

This PR add a new strategy, when `limit` is small and some string columns need to be materialized, we fallback to origin way to sort small blocks to avoid `merge`.

I tested on ckbench and find out 256 is a reasonable threshold to do this fallback.

# Proposed changes

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/exec/vsort_node.cpp [hunk 1]
```diff
@@ -182,6 +199,56 @@ Status VSortNode::sort_input(RuntimeState* state) {
...
+                            new SortCursorImpl(_sorted_blocks.back(), _sort_description)));
+                } else {
+                    SortBlockCursor block_cursor(
+                            _pool->add(new SortCursorImpl(block, _sort_description)));
+                    if (!block_cursor.totally_greater(_block_priority_queue.top())) {
+                        _sorted_blocks.emplace_back(std::move(block));
+                        _block_priority_queue.push(block_cursor);
+                        _total_mem_usage += mem_usage;
+                    } else {
+                        continue;
```
Comment (HappenLee) [inline be/src/vec/exec/vsort_node.cpp:234]: maybe in the case we should reuse the block mem?

File: be/src/vec/exec/vsort_node.cpp [hunk 2]
```diff
@@ -182,6 +199,56 @@ Status VSortNode::sort_input(RuntimeState* state) {
...
+                    }
+                }
+            } else {
+                // dispose normal sort logic
+                _total_mem_usage += mem_usage;
```
Comment (HappenLee) [inline be/src/vec/exec/vsort_node.cpp:239]: why row == 0 do the work ?

Comment (Gabriel39) [reply be/src/vec/exec/vsort_node.cpp:239]: `_limit == -1` instead of `row == 0` here

File: be/src/vec/common/sort/sorter.cpp [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+            *mem_reuse = false;
+        } else {
+            Block sorted_block;
+            sorted_block.swap(*block);
+            SortBlockCursor block_cursor(
+                    _pool->add(new SortCursorImpl(sorted_block, _sort_description)));
+            if (!block_cursor.totally_greater(_block_priority_queue.top())) {
+                _state->sorted_blocks.emplace_back(std::move(sorted_block));
+                _block_priority_queue.push(block_cursor);
+                *mem_reuse = false;
```
Comment (HappenLee) [inline be/src/vec/common/sort/sorter.cpp:236]: you can set false in the first line of the function. only need set `true` else

File: be/src/vec/common/sort/sorter.cpp [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+    RETURN_IF_ERROR(_do_sort(block, mem_reuse));
+    return Status::OK();
+}
+
+Status TopNSorter::prepare_for_read() {
+    _state->build_merge_tree(_sort_description);
+    return Status::OK();
+}
+
+Status TopNSorter::get_next(RuntimeState* state, Block* block, bool* eos) {
```
Comment (HappenLee) [inline be/src/vec/common/sort/sorter.cpp:197]: seems the code is same as `FullSorter` `get_next` may should not virtual function and in `Sorter `

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**